### PR TITLE
Adds the possibility of using innerJoin after contain

### PR DESCRIPTION
**Pull request by DiogoRMax**
Example of use:

$Users->find()->contain(['Contacts'])->innerJoin(['CT' => 'contact_types'], ['CT.id = Contacts.contact_type_id']);

### CODE REVIEW

**Comment by inoas** [comment]
Most important would be unit tests that proof that this did not work before and that proof that it works afterwards I suppose as changes to the ORM can break real world apps quite easily.

**Comment by DiogoRMax** [comment]
Excuses I do not know how to do these unit tests.

However the sql generated when using the current ORM is like:

> SELECT * 
> FROM users Users
> 	LEFT JOIN contact_types CT ON (
> 		CT.id = Contacts.contact_type_id
> 	)
> 	LEFT JOIN contacts Contacts ON (
> 		Contacts.user_id = (Users.id)
> 	)

What causes error

With the change it puts the innerJoin afterwards like this:

> SELECT * 
> FROM users Users
> 	LEFT JOIN contacts Contacts ON (
> 		Contacts.user_id = (Users.id)
> 	)
> 	LEFT JOIN contact_types CT ON (
> 		CT.id = Contacts.contact_type_id
> 	)


If you can help me with the process, I'm grateful ....

**Comment by markstory** [comment]
Your changes [cause existing tests to fail](https://travis-ci.org/cakephp/cakephp/jobs/317401087) so there is likely a regression against existing behavior being introduced.

File: src/ORM/Query.php [hunk 1]
```diff
@@ -1059,12 +1059,20 @@ protected function _transformQuery()
...
         if (empty($this->_parts['from'])) {
             $this->from([$this->_repository->getAlias() => $this->_repository->table()]);
         }
         $this->_addDefaultFields();
         $this->getEagerLoader()->attachAssociations($this, $this->_repository, !$this->_hasFields);
         $this->_addDefaultSelectTypes();
+
+        $i = count($this->_parts['join']);
+        foreach ($joins as $a => $j) {
+            $this->_parts['join'][is_string($a) ? $a : $i++] = $j;
```
Comment (markstory) [inline src/ORM/Query.php]: Couldn't this result in joins created by the eagerloader being overwritten?

Comment (DiogoRMax) [reply src/ORM/Query.php]: The problem is in the order in which JOINS are created.

The eagerloader is processed last.
Indepedent order to call the method.

Comment (DiogoRMax) [reply src/ORM/Query.php]: The problem is in the order in which JOINS are created.

The eagerloader is processed last.
Indepedent order to call the method.

**Comment by DiogoRMax** [comment]
The problem is in the order in which JOINS are created.

The eagerloader is processed last.
Indepedent order to call the method.

**Comment by DiogoRMax** [comment]
Is there no way that the call order can be respected?

**Comment by DiogoRMax** [comment]
I believe adding the respect to the invocation order of the method makes the ORM much more powerful and versatile.

But I also agree that it may end up breaking some apps.

The alternative would be to create a smart computer, which takes into account the dependency of JOINS.

If this change can be accepted, I can create the code.

**Comment by inoas** [comment]
Are you certain the order of non-nested LEFT-JOINS matters as shown in your examples? As far as I know it does not.

**Comment by lorenzo** [comment]
Have you tried using `matching()` or `innerJoinWith()`? I believe either of those will solve your problem

**Comment by DiogoRMax** [comment]
Sorry for the inconvenience, I'm new to this business.

**Comment by DiogoRMax** [comment]
How can errors occur in the tests, and the code added, should not change the flow of the program in any way?

File: src/Database/Query.php [hunk 2]
```diff
@@ -1926,6 +1934,31 @@ public function isBufferedResultsEnabled()
...
+     */
+    public function enableSmartJoin($enable = true)
+    {
+        $this->_smartJoin = (bool)$enable;
+
+        return $this;
+    }
+
+    /**
+     * Returns the current hydration mode.
```
Comment (markstory) [inline src/Database/Query.php]: Documentation needs to be updated.

File: src/Database/QueryCompiler.php [hunk 3]
```diff
@@ -250,6 +255,71 @@ protected function _buildJoinPart($parts, $query, $generator)
...
+            $alias[trim($join['alias'], '`')] = $join['alias'];
+            $alias[trim($join['table'], '`')] = $join['alias'];
+        }
+
+        $part = '';
+        foreach ($order as $alias) {
+            $part .= $joins[$alias];
+        }
+
+        return $part;
```
Comment (markstory) [inline src/Database/QueryCompiler.php]: This is a significant amount of code without any tests. Adding new modes increases complexity quite a bit as this mode can be combined with every other possible state. Is there a way to accomplish what you are trying to do without additional modes?

Comment (DiogoRMax) [reply src/Database/QueryCompiler.php]: I can not imagine any solution that will not break applications already deployed in the real world.

My point is:

$query->innerJoin(A...)->innerJoin(B...)
does not generate the same SQL as:
$query->innerJoin(B...)->innerJoin(A...)

But ...

$query->contain(A...)->innerJoin(B...)
$query->innerJoin(B...)->contain(A...)
Generates the same SQL, always leaving the JOIN part of the 'contain' at the end.

Comment (DiogoRMax) [reply src/Database/QueryCompiler.php]: Has anyone ever had this difficulty?

The only way I can see not to break the implementations already made, is to create an intelligent ordering system for the JOIN part.

Comment (DiogoRMax) [reply src/Database/QueryCompiler.php]: Or maybe create some parameter that holds the order of the containsJOINs

Comment (markstory) [reply src/Database/QueryCompiler.php]: Have you tried replacing `innerJoin` with `innerJoinWith` as @lorenzo suggested earlier? `innerJoinWith` uses the eagerloader so it should preserve join ordering.

> Has anyone ever had this difficulty?

Yes a very similar issue was reported in #10746 but there are a few solutions to that problem.

**Comment by DiogoRMax** [comment]
Thank you...
But my project does not understand the use of innerJoinWith, and thinking about it is a particularity that does not have to be included in Cakephp.

I have solved my problems with other solutions, thanks for the attention ...

**Pull request closed without merging**